tools/docker/static-analysis/cppcheck.sh

```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+
+scriptdir="$(cd "${0%/*}"; pwd)"
+rootdir="${scriptdir%/*/*/*}"
+
+. "${rootdir}/tools/docker/functions.sh"
+
+OUTPUT_FILE="$rootdir/cppcheck_results.txt"
+INCLUDES="\
+-I$rootdir/framework/platform/bpl/include \
+-I$rootdir/framework/transport/ieee1905_transport/include \
+-I$rootdir/framework/common/include \
+-I$rootdir/framework/tlvf/src/include \
+-I$rootdir/framework/tlvf/AutoGenerated/include \
+-I$rootdir/framework/discovery_agent/include \
+-I$rootdir/common/beerocks/btl/include \
+-I$rootdir/common/beerocks/tlvf/src/include \
+-I$rootdir/common/beerocks/tlvf/AutoGenerated/include \
+-I$rootdir/common/beerocks/bcl/include \
+-I$rootdir/common/beerocks/bwl/include"
+
+usage() {
+    echo "usage: $(basename "$0") <source> [source]"
+    echo "run cpp check on <source> and output number of issues per severity"
+}
+
+run_cppcheck() {
+    # --force is needed to check more than 12 ifdefs
+    cppcheck --template='{file}:{line}:{column}: {severity}: {message} [{id}]\n{code}' --force --error-exitcode=1 --enable=warning,style,information -q $INCLUDES "$@" 2>&1 | tee "$OUTPUT_FILE"
+}
+
+colorize_severity() {
+    local severity msg
+    severity="$1"
+    msg="$2"
+    case "$severity" in
+        error)
+            printf '\033[1;31m%s\033[0m\n' "$msg"
+            ;;
+        warning)
+            printf '\033[1;33m%s\033[0m\n' "$msg"
+            ;;
+        style)
+            printf '\033[1;34m%s\033[0m\n' "$msg"
+            ;;
+        performance)
+            printf '\033[1;35m%s\033[0m\n' "$msg"
+            ;;
+        portability)
+            printf '\033[1;36m%s\033[0m\n' "$msg"
+            ;;
+        ok)
+            printf '\033[1;90m%s\033[0m\n' "$msg"
+            ;;
+
+        *)
+            echo unknown severity "$severity"
+            exit 1
+            ;;
+    esac
+}
+
+output_results() {
+    local severities status
+    severities="error warning style performance portability"
+    status=0
+    for s in $severities; do
+        nb_issues="$(grep -Ec "^[^ ]+ $s:" "$OUTPUT_FILE")"
+        if [ $nb_issues -gt 0 ] ; then
+            colorize_severity "$s" "$nb_issues issues with severity $s"
+            case "$s" in
+                # report the following severities in exit code:
+                error|warning)
+                    status=$((status + 1))
```

Comment:
Note that if we have exactly 256 errors+warnings, this will overflow back to 0. Not important enough to fix IMO though.

Reply:
I think it is important. Glibc docs even explicitly warn against this:
```
Warning: Don’t try to use the number of errors as the exit status. This is actually not very useful; a parent process would generally not care how many errors occurred. Worse than that, it does not work, because the status value is truncated to eight bits. Thus, if the program tried to report 256 errors, the parent would receive a report of 0 errors—that is, success.

For the same reason, it does not work to use the value of errno as the exit status—these can exceed 255. 
```
https://www.gnu.org/software/libc/manual/html_node/Exit-Status.html

Reply:
Actually the exact value of the exit code is not used, only the fact that it's non-zero, so I'll just change it to `status=1`.